Recover cart after payment errors

kleinmann opened the pull request:
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfil our contribution guideline (https://docs.shopware.com/en/shopware-platform-dev-en/community/contribution-guideline?category=shopware-platform-dev-en/community).

Do your changes need to be mentioned in the documentation?
Add notes on your change right now in the documentation files in /src/Docs/Resources and add them to the pull request as well. 
-->

### 1. Why is this change necessary?
Currently the cart is lost if anything goes wrong during payments (be it actual errors or a customer cancellation on async payments). This PR recovers the cart from the created order. If the customer then retries the order, the previous order ID and order number will be transparently reused (thanks to the OrderConverter).

### 2. What does this change do, exactly?
It restores the cart from the cancelled order and redirects the customer back to the checkout (on Storefront sales channels).

### 3. Describe each step to reproduce the issue or behaviour.
* Add product to cart and go to checkout
* Force an error during payment (either by throwing it in DefaultPaymentHandler or cancelling e.g. in PayPal)
* Be sad because your cart is now gone

### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

bcremer commented:
@shopware/shopware-webdev can you please verify this?

klarstil commented:
Thanks for the PR, @benboit The code looks fine, I verified the behavior and the PR makes sense :) @bcremer Would you be so kind to merge the PR?

bcremer commented:
Thank you @benboit :+1:

kleinmann committed:
Recover an order's cart to the customer on payment errors.

kleinmann committed:
Upsert orders instead of just creating them.

Recovered carts retain the original order ID and order number. The
EntityWriter would try to update the database row, but due to the create
call the command will asserted as an InsertCommand instead of the
resulting UpdateCommand. By upserting here, this should handle both
cases (new carts and recovered carts).

kleinmann committed:
Fix tests using addresses and shipping method.

OliverSkroblin commented:
Hey @kleinmann,

great work! Could you provide some unit tests for this behavior?

kleinmann committed:
Add tests for cart recovery on payment errors.

kleinmann commented:
@OliverSkroblin I added two tests for recovery on payment errors :-)

OliverSkroblin commented:
Hi @kleinmann,

I imported your pull request and it's being tested off the pipeline right now. We are still discussing whether it will be merged into the 6.1.1 version or whether it will only be available from 6.2.

OliverSkroblin commented:
Hi @kleinmann,

as discussed on the phone, I close the pull request. As a background for other people who found this pull request:

It is not intended that the same order is placed several times via the storefront. If an error occurs in the payment process, it should be possible to restart this process without having to repeat the order process.
The order is already registered in the system and all business events have expired (stock refresh, mails, ...).

We have already planned to redirect the customer, in the future, to his account where he can resume the payment process with another, or the same, payment method. Due to a lack of resources we have unfortunately not been able to do so yet.
